File: languages/java/runtime/src/main/java/co/revaly/sdk/reconcile/RapReconciler.java
package co.revaly.sdk.reconcile;

import co.revaly.sdk.core.model.PendingTransactionResponse;
import co.revaly.sdk.core.model.TransactionResponse;
import co.revaly.sdk.errors.FailureClassifier;
import co.revaly.sdk.errors.RapCoreException;
import co.revaly.sdk.errors.RapFailureClass;
import co.revaly.sdk.logging.RapScrubber;
import co.revaly.sdk.logging.RapWireTrace;
import co.revaly.sdk.transport.RapHeaders;
import co.revaly.sdk.transport.RapRequestDecorator;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.io.IOException;
import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.function.Consumer;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * The OutcomeUnknown reconciliation procedure (failover-contract §3): GET-only, side-effect-free,
 * caller-bounded — the only loop the runtime owns (ADR-SDK-004).
 *
 * <p>This helper issues the merchant-transaction GET at the raw HTTP level (through the same shared
 * {@link HttpClient} and header decoration as every other SDK request) instead of the generated
 * core binding, for a safety reason: the core's generated oneOf wrapper for this endpoint requires
 * exactly one schema match, but the response schemas are all-optional under Jackson's lenient
 * binding, so valid 200 bodies can multi-match and fail deserialization inside the core (defect
 * flagged for an upstream/template fix — see the PR notes). Classification here therefore works
 * from the RAW body: the required {@code state} field discriminates a pending intent (it exists
 * only on the pending schema), and terminal records bind directly to {@link TransactionResponse} —
 * classify from raw bodies, never core wrappers (repo rule 5).
 */
public final class RapReconciler {

    private static final Logger LOG = LoggerFactory.getLogger(RapReconciler.class);

    private final HttpClient httpClient;
    private final String baseUri;
    private final RapRequestDecorator decorator;
    private final Duration overallDeadline;
    private final String apiVersion;
    private final ObjectMapper coreMapper;
    private final Consumer<RapWireTrace> wireTraceHook;

    public RapReconciler(
            HttpClient httpClient,
            String baseUri,
            RapRequestDecorator decorator,
            Duration overallDeadline,
            String apiVersion,
            ObjectMapper coreMapper,
            Consumer<RapWireTrace> wireTraceHook) {
        this.httpClient = httpClient;
        this.baseUri = baseUri.endsWith("/") ? baseUri.substring(0, baseUri.length() - 1) : baseUri;
        this.decorator = decorator;
        this.overallDeadline = overallDeadline;
        this.apiVersion = apiVersion;
        this.coreMapper = coreMapper;
        this.wireTraceHook = wireTraceHook;
    }

    /**
     * Runs the reconcile loop until a record is visible or the policy bounds are spent.
     *
     * @throws RapCoreException only for a rejected READ that polling can never fix
     *     (PermanentRejection other than 404 — bad credentials, malformed id); 404 is the
     *     NotFoundYet signal, and degraded reads (5xx/timeouts/transport) keep polling within the
     *     budget.
     * @throws InterruptedException when the calling thread is interrupted — the java cancellation
     *     idiom for a blocking helper; no verdict is implied.
     */
    public RapReconcileVerdict reconcile(String merchantTransactionId, ReconcilePolicy policy)
            throws RapCoreException, InterruptedException {
        if (merchantTransactionId == null || merchantTransactionId.trim().isEmpty()) {
            throw new IllegalArgumentException("merchantTransactionId is required");
        }
        if (policy == null) {
            throw new IllegalArgumentException("policy is required");
        }

        String path = "/transactions/merchant/" + urlEncode(merchantTransactionId);
        long startNanos = System.nanoTime();
        int attempts = 0;
        String lastCorrelationId = null;
        Integer lastHttpStatus = null;

        while (true) {
            attempts++;

            HttpResponse<String> response = null;
            try {
                response = send(path);
            } catch (IOException e) {
                // Transport failure on the READ: never-sent proof or not, the WRITE's
                // status is still unknown — keep polling within the caller's budget.
                RapCoreException classified = FailureClassifier.classifyTransportFailure(e);
                LOG.warn(
                        "rap.reconcile attempt {} transport failure ({}); continuing within policy",
                        attempts,
                        classified.getFailureClass());
                trace(path, null, null, null);
            }

            if (response != null) {
                lastHttpStatus = response.statusCode();
                String correlationId =
                        response.headers().firstValue(RapHeaders.CORRELATION_ID).orElse(null);
                if (correlationId != null) {
                    lastCorrelationId = correlationId;
                }
                String body = response.body();
                trace(path, response.statusCode(), correlationId, body);

                if (response.statusCode() / 100 == 2) {
                    RapReconcileVerdict found = readFound(body, lastCorrelationId, attempts);
                    if (found != null) {
                        return found;
                    }
                    // 2xx that did not parse: ambiguous read — poll again within budget.
                } else if (response.statusCode() == 404) {
                    // Not yet visible — the NotFoundYet signal, not an error (§3).
                    LOG.debug("rap.reconcile attempt {} not visible yet (404)", attempts);
                } else {
                    RapCoreException classified =
                            FailureClassifier.classifyResponse(
                                    response.statusCode(), body, apiVersion, correlationId);
                    if (classified.getFailureClass() == RapFailureClass.PERMANENT_REJECTION) {
                        // 400/401/403/422 escape: polling will never fix a rejected read
                        // (bad credentials, malformed id) — the caller must see it.
                        throw classified;
                    }
                    // Degraded read path (5xx/timeout on the GET) — exactly the window
                    // where visibility is widest; keep polling within the budget.
                    LOG.warn(
                            "rap.reconcile attempt {} degraded read [status={} class={}]; continuing within policy",
                            attempts,
                            response.statusCode(),
                            classified.getFailureClass());
                }
            }

            if (attempts >= policy.getMaxAttempts()) {
                break;
            }

            Duration elapsed = Duration.ofNanos(System.nanoTime() - startNanos);
            Duration delay = policy.delayForAttempt(attempts);
            if (elapsed.plus(delay).compareTo(policy.getOverallBudget()) >= 0) {
                break;
            }

            if (!delay.isZero()) {
                Thread.sleep(delay.toMillis());
            }
        }

        Duration elapsed = Duration.ofNanos(System.nanoTime() - startNanos);
        LOG.info(
                "rap.reconcile verdict=NotFoundYet attempts={} elapsedMs={} lastStatus={} correlation={}",
                attempts,
                elapsed.toMillis(),
                lastHttpStatus,
                lastCorrelationId);
        return new RapReconcileVerdict.NotFoundYet(
                attempts, elapsed, lastCorrelationId, lastHttpStatus);
    }

    private HttpResponse<String> send(String path) throws IOException, InterruptedException {
        HttpRequest.Builder builder =
                HttpRequest.newBuilder()
                        .uri(URI.create(baseUri + path))
                        .header("Accept", "application/json")
                        .GET();
        if (overallDeadline != null) {
            builder.timeout(overallDeadline);
        }
        decorator.accept(builder);
        return httpClient.send(builder.build(), HttpResponse.BodyHandlers.ofString());
    }

    /**
     * Maps a 2xx body to a Found verdict from the RAW json. Returns null for a body this SDK cannot
     * read at all (→ poll-continue: an ambiguous read is not a sighting).
     */
    private RapReconcileVerdict readFound(String body, String correlationId, int attempt) {
        JsonNode root;
        try {
            root = coreMapper.readTree(body == null ? "" : body);
        } catch (IOException e) {
            root = null;
        }
        if (root == null || !root.isObject()) {
            LOG.warn(
                    "rap.reconcile attempt {} returned 2xx with an unreadable body; continuing within policy",
                    attempt);
            return null;
        }

        // `state` exists only on the pending schema — its presence is authoritative
        // (the spec marks it the discriminator).
        JsonNode state = root.get("state");
        if (state != null && state.isTextual()) {
            PendingTransactionResponse pending = null;
            try {
                pending = coreMapper.treeToValue(root, PendingTransactionResponse.class);
            } catch (IOException e) {
                // A pending-shaped record this SDK version cannot bind is still a
                // sighting — surface it conservatively rather than polling on.
            }
            if (pending != null) {
                LOG.info(
                        "rap.reconcile verdict=Found outcome=PENDING correlation={}",
                        correlationId);
                return new RapReconcileVerdict.Found(
                        RapTransactionOutcome.PENDING, null, pending, correlationId);
            }
            return new RapReconcileVerdict.Found(
                    RapTransactionOutcome.UNKNOWN, null, null, correlationId);
        }

        // Terminal records bind DIRECTLY to TransactionResponse — never through the
        // core's oneOf wrapper (see the class doc).
        try {
            TransactionResponse transaction =
                    coreMapper.treeToValue(root, TransactionResponse.class);
            RapTransactionOutcome outcome = mapOutcome(transaction.getTransactionStatus());
            LOG.info(
                    "rap.reconcile verdict=Found outcome={} correlation={}",
                    outcome,
                    correlationId);
            return new RapReconcileVerdict.Found(outcome, transaction, null, correlationId);
        } catch (IOException e) {
            // A response shape this SDK version does not recognize (e.g. a grouped
            // envelope, or a post-P-2 variant). Found-but-unmapped is still FOUND.
            return new RapReconcileVerdict.Found(
                    RapTransactionOutcome.UNKNOWN, null, null, correlationId);
        }
    }

    private static RapTransactionOutcome mapOutcome(Integer transactionStatus) {
        if (transactionStatus == null) {
            return RapTransactionOutcome.UNKNOWN;
        }
        switch (transactionStatus) {
            case 1:
                return RapTransactionOutcome.APPROVED;
            case 2:
                return RapTransactionOutcome.DECLINED;
            case 3:
                return RapTransactionOutcome.ERROR;
            default:
                return RapTransactionOutcome.UNKNOWN;
        }
    }

    private void trace(String path, Integer status, String correlationId, String rawResponseBody) {
        if (wireTraceHook == null) {
            return;
        }
        try {
            wireTraceHook.accept(
                    new RapWireTrace(
                            "reconcile",
                            "GET",
                            path,
                            status,
                            correlationId,
                            null,
                            null,
                            null,
                            rawResponseBody == null
                                    ? null
                                    : RapScrubber.scrubJson(rawResponseBody)));
        } catch (RuntimeException e) {
            // Observer exceptions are swallowed (runtime-tdd §6) — tracing must never
            // change payment control flow.
            LOG.debug("rap.wiretrace hook threw; ignored", e);
        }
    }

    private static String urlEncode(String value) {
        return URLEncoder.encode(value, StandardCharsets.UTF_8).replace("+", "%20");
    }
}

File: languages/php/runtime/src/Reconcile/RapReconciler.php
<?php

declare(strict_types=1);

namespace Revaly\Sdk\Reconcile;

use GuzzleHttp\ClientInterface;
use GuzzleHttp\Psr7\Request;
use GuzzleHttp\RequestOptions;
use Psr\Http\Message\ResponseInterface;
use Psr\Log\LoggerInterface;
use Revaly\Sdk\Core\Model\PendingTransactionResponse;
use Revaly\Sdk\Core\Model\TransactionResponse;
use Revaly\Sdk\Core\ObjectSerializer;
use Revaly\Sdk\Errors\FailureClassifier;
use Revaly\Sdk\Errors\RapCoreException;
use Revaly\Sdk\Errors\RapFailureClass;
use Revaly\Sdk\Logging\RapScrubber;
use Revaly\Sdk\Logging\RapWireTrace;
use Revaly\Sdk\Transport\RapHeaders;

/**
 * The OutcomeUnknown reconciliation procedure (failover-contract §3): GET-only,
 * side-effect-free, caller-bounded — the only loop the runtime owns (ADR-SDK-004).
 *
 * This helper issues the merchant-transaction GET at the raw HTTP level (through the
 * same shared Guzzle client and header middleware as every other SDK request) instead of
 * the generated core binding, for a safety reason: the core's generated wrapper for this
 * endpoint merges ALL response branches into one property blob with no discrimination
 * (`GetTransactionByMerchantTransactionId200Response` carries terminal, pending AND
 * group fields side by side). Classification here therefore works from the RAW body: the
 * required `state` field discriminates a pending intent (it exists only on the pending
 * schema), and terminal records bind directly to {@see TransactionResponse} — classify
 * from raw bodies, never core wrappers (repo rule 5).
 */
final class RapReconciler
{
    public function __construct(
        private readonly ClientInterface $client,
        private readonly string $baseUrl,
        private readonly string $apiVersion,
        private readonly LoggerInterface $logger,
        private readonly ?\Closure $wireTraceHook,
    ) {
    }

    /**
     * Runs the reconcile loop until a record is visible or the policy bounds are spent.
     *
     * @throws RapCoreException only for a rejected READ that polling can never fix
     *         (PermanentRejection other than 404 — bad credentials, malformed id); 404
     *         is the NotFoundYet signal, and degraded reads (5xx/timeouts/transport)
     *         keep polling within the budget.
     */
    public function reconcile(string $merchantTransactionId, ReconcilePolicy $policy): RapReconcileVerdict
    {
        if (trim($merchantTransactionId) === '') {
            throw new \InvalidArgumentException('merchantTransactionId is required');
        }

        $path = '/transactions/merchant/' . rawurlencode($merchantTransactionId);
        $start = microtime(true);
        $attempts = 0;
        $lastCorrelationId = null;
        $lastHttpStatus = null;

        while (true) {
            $attempts++;

            $response = null;
            try {
                $response = $this->client->send(
                    new Request('GET', $this->baseUrl . $path, ['Accept' => 'application/json']),
                    [RequestOptions::HTTP_ERRORS => false],
                );
            } catch (\Throwable $failure) {
                // Transport failure on the READ: never-sent proof or not, the WRITE's
                // status is still unknown — keep polling within the caller's budget.
                $classified = FailureClassifier::classifyTransportRejection($failure);
                $this->logger->warning(
                    'rap.reconcile attempt {attempt} transport failure ({class}); continuing within policy',
                    ['attempt' => $attempts, 'class' => $classified->getFailureClass()->value],
                );
                $this->trace($path, null, null, null);
            }

            if ($response !== null) {
                $lastHttpStatus = $response->getStatusCode();
                $correlationId = $response->getHeaderLine(RapHeaders::CORRELATION_ID);
                $correlationId = $correlationId === '' ? null : $correlationId;
                if ($correlationId !== null) {
                    $lastCorrelationId = $correlationId;
                }
                $body = (string) $response->getBody();
                $this->trace($path, $response->getStatusCode(), $correlationId, $body);

                if (intdiv($response->getStatusCode(), 100) === 2) {
                    $found = $this->readFound($body, $lastCorrelationId, $attempts);
                    if ($found !== null) {
                        return $found;
                    }
                    // 2xx that did not parse: ambiguous read — poll again within budget.
                } elseif ($response->getStatusCode() === 404) {
                    // Not yet visible — the NotFoundYet signal, not an error (§3).
                    $this->logger->debug(
                        'rap.reconcile attempt {attempt} not visible yet (404)',
                        ['attempt' => $attempts],
                    );
                } else {
                    $classified = FailureClassifier::classifyResponse(
                        $response->getStatusCode(),
                        $body,
                        $this->apiVersion,
                        $correlationId,
                    );
                    if ($classified->getFailureClass() === RapFailureClass::PermanentRejection) {
                        // 400/401/403/422 escape: polling will never fix a rejected read
                        // (bad credentials, malformed id) — the caller must see it.
                        throw $classified;
                    }
                    // Degraded read path (5xx/timeout on the GET) — exactly the window
                    // where visibility is widest; keep polling within the budget.
                    $this->logger->warning(
                        'rap.reconcile attempt {attempt} degraded read [status={status} class={class}]; continuing within policy',
                        [
                            'attempt' => $attempts,
                            'status' => $response->getStatusCode(),
                            'class' => $classified->getFailureClass()->value,
                        ],
                    );
                }
            }

            if ($attempts >= $policy->getMaxAttempts()) {
                break;
            }

            $elapsed = microtime(true) - $start;
            $delay = $policy->delayForAttempt($attempts);
            if ($elapsed + $delay >= $policy->getOverallBudgetSeconds()) {
                break;
            }

            if ($delay > 0) {
                usleep((int) round($delay * 1_000_000));
            }
        }

        $elapsed = microtime(true) - $start;
        $this->logger->info(
            'rap.reconcile verdict=NotFoundYet attempts={attempts} elapsedMs={elapsedMs} lastStatus={lastStatus} correlation={correlation}',
            [
                'attempts' => $attempts,
                'elapsedMs' => (int) round($elapsed * 1000),
                'lastStatus' => $lastHttpStatus,
                'correlation' => $lastCorrelationId,
            ],
        );

        return new NotFoundYet($attempts, $elapsed, $lastCorrelationId, $lastHttpStatus);
    }

    /**
     * Maps a 2xx body to a Found verdict from the RAW json. Returns null for a body
     * this SDK cannot read at all (→ poll-continue: an ambiguous read is not a
     * sighting).
     */
    private function readFound(string $body, ?string $correlationId, int $attempt): ?RapReconcileVerdict
    {
        $root = json_decode($body, false);
        if (!is_object($root)) {
            $this->logger->warning(
                'rap.reconcile attempt {attempt} returned 2xx with an unreadable body; continuing within policy',
                ['attempt' => $attempt],
            );

            return null;
        }

        // `state` exists only on the pending schema — its presence is authoritative
        // (the spec marks it the discriminator).
        if (isset($root->state) && is_string($root->state)) {
            $pending = null;
            try {
                $bound = ObjectSerializer::deserialize($root, PendingTransactionResponse::class);
                $pending = $bound instanceof PendingTransactionResponse ? $bound : null;
            } catch (\Throwable $bindFailure) {
                // A pending-shaped record this SDK version cannot bind is still a
                // sighting — surface it conservatively rather than polling on.
            }
            if ($pending !== null) {
                $this->logger->info(
                    'rap.reconcile verdict=Found outcome=Pending correlation={correlation}',
                    ['correlation' => $correlationId],
                );

                return new Found(RapTransactionOutcome::Pending, null, $pending, $correlationId);
            }

            return new Found(RapTransactionOutcome::Unknown, null, null, $correlationId);
        }

        // Terminal records bind DIRECTLY to TransactionResponse — never through the
        // core's merged wrapper (see the class doc).
        try {
            $transaction = ObjectSerializer::deserialize($root, TransactionResponse::class);
            if (!$transaction instanceof TransactionResponse) {
                return new Found(RapTransactionOutcome::Unknown, null, null, $correlationId);
            }
            $outcome = self::mapOutcome($transaction->getTransactionStatus());
            $this->logger->info(
                'rap.reconcile verdict=Found outcome={outcome} correlation={correlation}',
                ['outcome' => $outcome->value, 'correlation' => $correlationId],
            );

            return new Found($outcome, $transaction, null, $correlationId);
        } catch (\Throwable $bindFailure) {
            // A response shape this SDK version does not recognize (e.g. a grouped
            // envelope, a new enum value, or a post-P-2 variant). Found-but-unmapped
            // is still FOUND.
            return new Found(RapTransactionOutcome::Unknown, null, null, $correlationId);
        }
    }

    private static function mapOutcome(?int $transactionStatus): RapTransactionOutcome
    {
        return match ($transactionStatus) {
            1 => RapTransactionOutcome::Approved,
            2 => RapTransactionOutcome::Declined,
            3 => RapTransactionOutcome::Error,
            default => RapTransactionOutcome::Unknown,
        };
    }

    private function trace(string $path, ?int $status, ?string $correlationId, ?string $rawResponseBody): void
    {
        if ($this->wireTraceHook === null) {
            return;
        }
        try {
            ($this->wireTraceHook)(new RapWireTrace(
                'reconcile',
                'GET',
                $path,
                $status,
                $correlationId,
                null,
                null,
                null,
                $rawResponseBody === null ? null : RapScrubber::scrubJson($rawResponseBody),
            ));
        } catch (\Throwable $hookFailure) {
            // Observer exceptions are swallowed (runtime-tdd §6) — tracing must never
            // change payment control flow.
            $this->logger->debug('rap.wiretrace hook threw; ignored', ['exception' => get_class($hookFailure)]);
        }
    }
}
